### PR TITLE
semantic compiler: Add analyzer struct

### PR DESCRIPTION
**Pull request by mattnibs**
Add central analyzer struct for semantic compilation.

### CODE REVIEW

File: compiler/semantic/ast.go [hunk 1]
```diff
@@ -13,5 +13,6 @@ import (
 // to DAG form, resolving syntax ambiguities, and performing constant propagation.
 // After semantic analysis, the DAG is ready for either optimization or compilation.
 func Analyze(ctx context.Context, seq ast.Seq, source *data.Source, head *lakeparse.Commitish) (dag.Seq, error) {
-	return semSeq(ctx, NewScope(), seq, source, head)
+	c := newChecker(ctx, source, head)
+	return c.check(seq)
```
Comment (mattnibs) [inline compiler/semantic/ast.go:17]: In the forthcoming user ops pr this will have additional steps.

File: compiler/semantic/op.go [hunk 2]
```diff
@@ -144,17 +144,17 @@ func semSortKey(p *ast.SortKey) (order.SortKey, error) {
...
+func (c *checker) semPool(scope *Scope, p *ast.Pool) ([]dag.Op, error) {
 	var poolNames []string
 	var err error
 	switch specPool := p.Spec.Pool.(type) {
 	case nil:
 		// This is a lake meta-query.
 		poolNames = []string{""}
 	case *ast.Glob:
-		poolNames, err = matchPools(ctx, ds, reglob.Reglob(specPool.Pattern), specPool.Pattern, "glob")
+		poolNames, err = matchPools(c.ctx, c.source, reglob.Reglob(specPool.Pattern), specPool.Pattern, "glob")
```
Comment (nwt) [inline compiler/semantic/op.go:155]: Make this a method so you don't have to pass c.ctx and c.source?

File: compiler/semantic/op.go [hunk 3]
```diff
@@ -165,7 +165,7 @@ func semPool(ctx context.Context, scope *Scope, p *ast.Pool, ds *data.Source, he
 	}
 	var sources []dag.Op
 	for _, name := range poolNames {
-		source, err := semPoolWithName(ctx, scope, p, name, ds, head)
+		source, err := semPoolWithName(c.ctx, scope, p, name, c.source, c.head)
```
Comment (nwt) [inline compiler/semantic/op.go:168]: Make this a checker method so you don't have to pass c.ctx, c.source, and c.head?

File: compiler/semantic/expr.go [hunk 4]
```diff
@@ -15,7 +15,7 @@ import (
 	"github.com/brimdata/zed/zson"
 )
 
-func semExpr(scope *Scope, e ast.Expr) (dag.Expr, error) {
+func (c *checker) semExpr(scope *Scope, e ast.Expr) (dag.Expr, error) {
```
Comment (nwt) [inline compiler/semantic/expr.go:18]: Any reason not to drop the scope parameter everywhere in favor of checker.scope?

Comment (mattnibs) [reply compiler/semantic/expr.go:18]: Yeah I was just thinking the same thing, I'll do this.

File: compiler/semantic/ast.go (hunk 1 above)
Comment (nwt) [inline compiler/semantic/ast.go:17]: Existing comments and error strings almost always refer to this step as semantic analysis, so should we rename "checker" to "analyzer" for consistency with that?

Comment (mattnibs) [reply compiler/semantic/ast.go:17]: ok

File: compiler/semantic/op.go [hunk 5]
```diff
@@ -931,9 +928,9 @@ func semCallOp(scope *Scope, call *ast.Call, seq dag.Seq) (dag.Seq, error) {
 	if !function.HasBoolResult(call.Name) {
 		return nil, nil
 	}
-	c, err := semCall(scope, call)
+	out, err := a.semCall(call)
```
Comment (nwt) [inline compiler/semantic/op.go:931]: Nit: Leave this name unchanged.
```suggestion
	c, err := a.semCall(call)
```

**Review by nwt — APPROVED** [review]

**Pull request merged**